README.rst
SMSC
====

.. image:: https://img.shields.io/pypi/v/smsc.svg
    :target: https://pypi.python.org/pypi/smsc

.. image:: https://readthedocs.org/projects/smsc_python/badge/?version=latest
    :target: http://smsc_python.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://travis-ci.org/Otetz/smsc.svg?branch=master
    :target: https://travis-ci.org/Otetz/smsc

.. image:: https://coveralls.io/repos/github/Otetz/smsc/badge.svg?branch=master
    :target: https://coveralls.io/github/Otetz/smsc?branch=master

.. image:: https://img.shields.io/codeclimate/github/Otetz/smsc.svg
    :target: https://codeclimate.com/github/Otetz/smsc

.. image:: https://img.shields.io/pypi/l/smsc.svg
    :target: https://pypi.python.org/pypi/smsc

.. image:: https://img.shields.io/pypi/pyversions/smsc.svg
    :target: https://pypi.python.org/pypi/smsc

SMSC.ru HTTP API Library.

Installation
------------

Install smsc package from `PyPI <https://pypi.python.org/pypi>`_:

.. code-block:: bash

    $ pip install smsc

Getting started
---------------

Basic usage sample:

.. code-block:: python

    >>> from smsc.messages import SMSMessage
    >>> from smsc.api import SMSC
    >>> client = SMSC(login='alexey', password='psw')
    >>> res = client.send(to='79999999999', message=SMSMessage(text='Hello, World!'))
    >>> res.count
    1
    >>> res.cost
    1.44

Documentation
-------------

Documentation is available at `Read the Docs <http://smsc_python.readthedocs.io/en/latest/>`_.

Links
-----

- `SMSC.ru HTTP API <https://smsc.ru/api/http/#menu>`_


HISTORY.rst
.. :changelog:

Release History
---------------

0.1.1 (2017-05-30)
++++++++++++++++++

* Default charset set to UTF-8.

0.1.0 (2017-05-29)
++++++++++++++++++

* Birth!


tests/test_smsc.py
# flake8: noqa

import datetime
import logging

import os
import pytest
import requests_mock
from furl import furl

from smsc import (
    SMSC, SMSMessage, SendResponse, CostResponse, StatusResponse, BalanceResponse, ViberMessage, FlashMessage, Status,
    SendError, GetCostError, GetStatusError, GetBalanceError
)

logging.basicConfig(level=logging.DEBUG, format="%(asctime)s {%(pathname)s:%(lineno)d} %(levelname)s - %(message)s")
log = logging.getLogger(__name__)
URL = 'https://smsc.ru/sys/'


@pytest.fixture
def client():
    return SMSC(login=os.environ['SMSC_LOGIN'], password=os.environ['SMSC_PASSWORD'],
                sender=os.environ.get('SMSC_SENDER', None))


@pytest.fixture
def params():
    return {'login': os.environ['SMSC_LOGIN'], 'psw': os.environ['SMSC_PASSWORD'],
            'sender': os.environ.get('SMSC_SENDER', None)}


def test_client():
    assert os.environ.get('SMSC_LOGIN')
    assert os.environ.get('SMSC_PASSWORD')
    client2 = SMSC(login=os.environ['SMSC_LOGIN'], password=os.environ['SMSC_PASSWORD'],
                   sender=os.environ.get('SMSC_SENDER', None))
    assert str(client2) == "<SMSC login='%s' sender='%s'>" % (
        os.environ['SMSC_LOGIN'], os.environ.get('SMSC_SENDER', None))


def test_message_sms():
    message = SMSMessage(text='test')
    assert message is not None
    assert isinstance(message, SMSMessage)
    assert str(message) == "<SMSMessage text=test format=None>"
    assert message.text == 'test'
    assert message.format is None
    enc = message.encode()
    assert isinstance(enc, dict)
    assert 'mes' in enc
    assert enc['mes'] is not None
    assert enc['mes'] == 'test'
    assert enc['charset'] is not None
    assert enc['charset'] == 'utf-8'
    assert len(enc.keys()) == 2


# noinspection PyShadowingNames
def test_sms_simple(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    message_text = 'test'
    f = furl(URL).add(path="send.php").add(params).add({'phones': to, 'mes': message_text, 'cost': 2})
    with requests_mock.Mocker() as m:
        m.get(f.url, json={'cnt': 1, 'id': 1, 'cost': 1.44},
              headers={'Content-Type': 'application/json; charset=utf-8'})
        res = client.send(to=os.environ['PHONE'], message=SMSMessage(text='test'))
    assert res is not None
    assert isinstance(res, SendResponse)
    assert str(res) == "<SendResponse id=1 count=1 cost=1.44>"
    assert res.message_id == 1
    assert res.count == 1
    assert res.cost == 1.44


# noinspection PyShadowingNames
def test_sms_simple_fail(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    message_text = 'test'
    f = furl(URL).add(path="send.php").add(params).add({'phones': to, 'mes': message_text, 'cost': 2})
    with requests_mock.Mocker() as m:
        m.get(f.url, json={}, headers={'Content-Type': 'application/json; charset=utf-8'}, status_code=404)
        with pytest.raises(SendError):
            client.send(to=os.environ['PHONE'], message=SMSMessage(text='test'))


# noinspection PyShadowingNames
def test_sms_cost(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    message_text = 'test'
    f = furl(URL).add(path="send.php").add(params).add({'phones': to, 'mes': message_text, 'cost': 1})
    with requests_mock.Mocker() as m:
        m.get(f.url, json={'cnt': 1, 'cost': 1.44}, headers={'Content-Type': 'application/json; charset=utf-8'})
        res = client.get_cost(to=os.environ['PHONE'], message=SMSMessage(text='test'))
    assert res is not None
    assert isinstance(res, CostResponse)
    assert str(res) == "<CostResponse count=1 cost=1.44>"
    assert res.count == 1
    assert res.cost == 1.44


# noinspection PyShadowingNames
def test_sms_cost_fail(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    message_text = 'test'
    f = furl(URL).add(path="send.php").add(params).add({'phones': to, 'mes': message_text, 'cost': 1})
    with requests_mock.Mocker() as m:
        m.get(f.url, json={}, headers={'Content-Type': 'application/json; charset=utf-8'}, status_code=404)
        with pytest.raises(GetCostError):
            client.get_cost(to=os.environ['PHONE'], message=SMSMessage(text='test'))


# noinspection PyShadowingNames
def test_sms_status(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    msg_id = '1'
    f = furl(URL).add(path="status.php").add(params).add(
        {'phone': to + ',', 'id': msg_id + ',', 'charset': 'utf-8', 'all': 2, 'fmt': 3}).remove('sender')
    with requests_mock.Mocker() as m:
        m.get(f.url, json=[
            {'id': 1, 'send_timestamp': 1495823967, 'message': 'test', 'status_name': 'Доставлено', 'cost': '1.20',
             'phone': '79262138080', 'sender_id': 'avto-disp', 'last_date': '26.05.2017 21:39:32',
             'region': 'г.Москва и Московская область', 'send_date': '26.05.2017 21:39:27',
             'last_timestamp': 1495823972, 'operator': 'МегаФон', 'status': 1, 'country': 'Россия'}],
              headers={'Content-Type': 'application/json; charset=utf-8'})
        res = client.get_status(to=os.environ['PHONE'], msg_id=msg_id)
    assert res is not None
    assert isinstance(res, list)
    assert len(res) == 1
    assert str(res).startswith('[<StatusResponse ')
    assert isinstance(res[0], StatusResponse)
    assert str(res[0]) == "<StatusResponse id=1 status= <Status status=1 name=Доставлено>>"
    assert res[0].status is not None
    assert isinstance(res[0].status, Status)
    assert res[0].status.status_id == 1
    assert res[0].status.name == 'Доставлено'
    assert res[0].data is not None
    assert isinstance(res[0].data, dict)
    for name in ['id', 'message', 'cost', 'phone', 'sender_id', 'last_date', 'region', 'send_date', 'operator',
                 'country']:
        assert name in res[0].data
    for name in ['send_timestamp', 'last_timestamp', 'status', 'status_name']:
        assert name not in res[0].data
    assert isinstance(res[0].data['send_date'], datetime.datetime)


# noinspection PyShadowingNames
def test_sms_status_fail(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    msg_id = '1'
    f = furl(URL).add(path="status.php").add(params).add(
        {'phone': to + ',', 'id': msg_id + ',', 'charset': 'utf-8', 'all': 2, 'fmt': 3}).remove('sender')
    with requests_mock.Mocker() as m:
        m.get(f.url, json={}, headers={'Content-Type': 'application/json; charset=utf-8'}, status_code=404)
        with pytest.raises(GetStatusError):
            client.get_status(to=os.environ['PHONE'], msg_id=msg_id)


# noinspection PyShadowingNames
def test_get_balance(client: SMSC, params: dict):
    f = furl(URL).add(path="balance.php").add(params).remove('sender')
    with requests_mock.Mocker() as m:
        m.get(f.url, json={'balance': '100.01', 'currency': 'RUR'},
              headers={'Content-Type': 'application/json; charset=utf-8'})
        res = client.get_balance()
    assert res is not None
    assert isinstance(res, BalanceResponse)
    assert str(res) == "<BalanceResponse balance=100.01 credit=0.00 currency=RUR>"
    assert res.balance == 100.01
    assert res.credit == 0.0
    assert res.currency == 'RUR'


# noinspection PyShadowingNames
def test_get_balance_fail(client: SMSC, params: dict):
    f = furl(URL).add(path="balance.php").add(params).remove('sender')
    with requests_mock.Mocker() as m:
        m.get(f.url, json={}, headers={'Content-Type': 'application/json; charset=utf-8'}, status_code=404)
        with pytest.raises(GetBalanceError):
            client.get_balance()


# noinspection PyShadowingNames
def test_viber(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    message_text = 'test'
    f = furl(URL).add(path="send.php").add(params).add({'phones': to, 'mes': message_text, 'cost': 2, 'viber': 1})
    with requests_mock.Mocker() as m:
        m.get(f.url, json={'error': "can't to deliver", 'id': 1, 'error_code': 8},
              headers={'Content-Type': 'application/json; charset=utf-8'})
        res = client.send(to=os.environ['PHONE'], message=ViberMessage(text='test'))
    assert res is not None
    assert isinstance(res, SendResponse)
    assert res.error is not None
    assert str(res.error) == 'SMSCError(code=8, error="can\'t to deliver")'


# noinspection PyShadowingNames
def test_flash(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    message_text = 'test'
    f = furl(URL).add(path="send.php").add(params).add({'phones': to, 'mes': message_text, 'cost': 2, 'flash': 1})
    with requests_mock.Mocker() as m:
        m.get(f.url, json={'cnt': 1, 'id': 1, 'cost': 1.44},
              headers={'Content-Type': 'application/json; charset=utf-8'})
        res = client.send(to=os.environ['PHONE'], message=FlashMessage(text='test'))
    assert res is not None
    assert isinstance(res, SendResponse)
    assert str(res) == "<SendResponse id=1 count=1 cost=1.44>"
    assert res.message_id == 1
    assert res.count == 1
    assert res.cost == 1.44


# noinspection PyShadowingNames
def test_sms_options(client: SMSC, params: dict):
    assert os.environ.get('PHONE')
    to = os.environ['PHONE']
    message_text = 'test'
    f = furl(URL).add(path="send.php").add(params).add(
        {'phones': to, 'mes': message_text, 'cost': 2, 'translit': 1, 'tinyurl': 1, 'maxsms': 1})
    with requests_mock.Mocker() as m:
        m.get(f.url, json={'cnt': 1, 'id': 1, 'cost': 1.44},
              headers={'Content-Type': 'application/json; charset=utf-8'})
        res = client.send(to=os.environ['PHONE'], message=SMSMessage(text='test', translit=1, tinyurl=1, maxsms=1))
    assert res is not None
    assert res.error is None
    assert isinstance(res, SendResponse)
    assert str(res) == "<SendResponse id=1 count=1 cost=1.44>"
    assert res.message_id == 1
    assert res.count == 1
    assert res.cost == 1.44


smsc/__init__.py
# -*- coding: utf-8 -*-
"""
SMSC.ru API Library.

Smsc is an library to send messages through SMSC.ru HTTP API.
Basic usage:

    >>> from smsc.messages import SMSMessage
    >>> from smsc.api import SMSC
    >>> client = SMSC(login='alexey', password='psw')
    >>> res = client.send(to='79999999999', message=SMSMessage(text='Hello, World!'))  # doctest: +SKIP
    >>> res.count # doctest: +SKIP
    1
    >>> res.cost # doctest: +SKIP
    1.44

The some other API methods are supported - see `smsc.api`. Full documentation is at
`Read the Docs <http://smsc_python.readthedocs.io/en/latest/>`_.

:copyright: (c) 2017 by Alexey Shevchenko.
:license: MIT, see LICENSE for more details.
"""

import logging  # noqa: T005
import sys  # noqa: T005

# noinspection PyUnresolvedReferences
from .api import SMSC  # noqa: F401
# noinspection PyUnresolvedReferences
from .responses import SendResponse, CostResponse, StatusResponse, BalanceResponse, Status  # noqa: F401
# noinspection PyUnresolvedReferences
from .messages import SMSMessage, FlashMessage, ViberMessage  # noqa: F401
# noinspection PyUnresolvedReferences
from .exceptions import SMSCException, SendError, GetCostError, GetBalanceError, GetStatusError  # noqa: F401

# Set default logging handler to avoid "No handler found" warnings.
try:  # Python 2.7+
    if sys.version_info >= (2, 7):
        from logging import NullHandler
except ImportError:
    if sys.version_info < (2, 7):
        class NullHandler(logging.Handler):
            """
            NullHandler implementation for Python < 2.7.

            This handler does nothing. It's intended to be used to avoid the
            "No handlers could be found for logger XXX" one-off warning. This is
            important for library code, which may contain code to log events. If a user
            of the library does not configure logging, the one-off warning might be
            produced; to avoid this, the library developer simply needs to instantiate
            a NullHandler and add it to the top-level logger of the library module or
            package.
            """

            def emit(self, record):
                """Stub."""

logging.getLogger(__name__).addHandler(NullHandler())


smsc/messages.py
# -*- coding: utf-8 -*-
"""
smsc.messages module.

This module contains the messages objects that power smsc library.

:copyright: (c) 2017 by Alexey Shevchenko.
:license: MIT, see LICENSE for more details.
"""
from abc import ABCMeta
from typing import Any, Dict, Optional


class Message(metaclass=ABCMeta):
    """
    Basic class for messages of ant type.

    Preferred for internal usage.

    :param str text: Text of the message
    :param Optional[str] msg_format: Message format. If None or empty - default, SMS message
    :param dict kwargs: Dictionary for optional API parameters
    """

    def __init__(self, text: str, msg_format: Optional[str] = None, **kwargs: dict) -> None:  # noqa: D102
        self._text = text
        self._format = msg_format
        self._translit = kwargs.get("translit")
        self._tinyurl = kwargs.get("tinyurl")
        self._maxsms = kwargs.get("maxsms")
        self._charset = "utf-8"

    @property
    def format(self) -> str:
        """Format of the message. Default is SMS Message."""
        return self._format

    @property
    def text(self) -> str:
        """Text of the message."""
        return self._text

    def encode(self) -> Dict[str, Any]:
        """Message parameters in dict, prepared for the API."""
        res = {"mes": self._text, "charset": self._charset}  # type: Dict[str, Any]
        if self._format is not None and self._format != "":
            res[self._format] = 1
        if self._translit is not None:
            res["translit"] = self._translit
        if self._tinyurl is not None:
            res["tinyurl"] = self._tinyurl
        if self._maxsms is not None:
            res["maxsms"] = self._maxsms
        return res

    def __str__(self) -> str:
        """Represent object as string."""
        return "<%s text=%s format=%s>" % (self.__class__.__name__, self._text, self._format)

    def __repr__(self) -> str:
        """Represent object for debug purposes."""
        return str(self)


class SMSMessage(Message):
    """
    SMS message type.

    :param str text: Text of the message
    :param dict kwargs: Dictionary for optional API parameters

    Usage::

        >>> from smsc import SMSMessage

        >>> m = SMSMessage(text="Hello, World!")
        >>> m
        <SMSMessage text=Hello, World! format=None>
    """

    def __init__(self, text: str, **kwargs: dict) -> None:  # noqa: D102
        assert len(text) <= 800
        super().__init__(text, msg_format=None, **kwargs)


class FlashMessage(Message):
    """
    Flash-SMS message type.

    :param str text: Text of the message
    :param dict kwargs: Dictionary for optional API parameters

    Usage::

        >>> from smsc import FlashMessage
        >>> m = FlashMessage(text="Hello, World!")
        >>> m
        <FlashMessage text=Hello, World! format=flash>
    """

    def __init__(self, text: str, **kwargs: dict) -> None:  # noqa: D102
        assert len(text) <= 800
        super().__init__(text, msg_format="flash", **kwargs)


class ViberMessage(Message):
    """
    Viber messenger message type.

    Note: Seems currently not working now.

    :param str text: Text of the message
    :param dict kwargs: Dictionary for optional API parameters

    Usage::

        >>> from smsc import ViberMessage
        >>> m = ViberMessage(text="Hello, World!")
        >>> m
        <ViberMessage text=Hello, World! format=viber>
    """

    def __init__(self, text: str, **kwargs: dict) -> None:  # noqa: D102
        assert len(text) <= 800
        super().__init__(text, msg_format="viber", **kwargs)


smsc/responses.py
# -*- coding: utf-8 -*-
"""
smsc.responses module.

This module contains the responses objects wrapping SMSC.ru API answers.

:copyright: (c) 2017 by Alexey Shevchenko.
:license: MIT, see LICENSE for more details.
"""
from abc import ABCMeta

from collections import namedtuple
from typing import Any, Dict, Optional

from arrow import get as arrow_get
from dateutil.tz import tz


# noinspection PyUnresolvedReferences
class SMSCError(namedtuple("SMSCError", ["code", "error"])):
    """
    Named tuple for error description with code.

    :param int code: Code of error
    :param str error: Description of error
    """


class Status:
    """
    Message delivery status with identification.

    :param dict obj: Dictionary from API JSON response
    """

    def __init__(self, obj: Dict[str, Any]) -> None:  # noqa: D102
        self.__id = int(obj.get("status", 0))
        self.__name = str(obj.get("status_name", ""))
        for name in ["status", "status_name"]:
            del obj[name]

    def __str__(self) -> str:
        """Represent object as string."""
        return "<%s status=%d name=%s>" % (self.__class__.__name__, self.__id, self.__name)

    def __repr__(self) -> str:
        """Represent object for debug purposes."""
        return str(self)

    @property
    def status_id(self) -> int:
        """Id of delivery status."""
        return self.__id

    @property
    def name(self) -> str:
        """Delivery status description."""
        return self.__name


class Response(metaclass=ABCMeta):
    """
    Basic class for response wrappers.

    :param dict obj: Dictionary from API JSON response
    """

    def __init__(self, obj: Dict[str, Any]) -> None:  # noqa: D102
        self.__error = obj.get("error", None)
        self.__code = int(obj.get("error_code", 0))

    @property
    def error(self) -> Optional[SMSCError]:
        """Error in response, if present."""
        if not self.__error and not self.__code:
            return None
        # noinspection PyTypeChecker
        return SMSCError(self.__code, self.__error)


class SendResponse(Response):
    """
    Response for send API command.

    :param dict obj: Dictionary from API JSON response
    """

    def __init__(self, obj: Dict[str, Any]) -> None:  # noqa: D102
        super().__init__(obj)
        self.__id = obj.get("id", None)
        self.__cnt = int(obj.get("cnt", 0))
        self.__cost = float(obj.get("cost", 0.0))

    def __str__(self) -> str:
        """Represent object as string."""
        return "<%s id=%s count=%d cost=%.2f>" % (
            self.__class__.__name__, self.__id, self.__cnt, self.__cost)

    def __repr__(self) -> str:
        """Represent object for debug purposes."""
        return str(self)

    @property
    def message_id(self) -> str:
        """Id of sent message."""
        return self.__id

    @property
    def count(self) -> int:
        """Count of billed message parts."""
        return self.__cnt

    @property
    def cost(self) -> float:
        """Cost of sent message."""
        return self.__cost


class CostResponse(Response):
    """
    Response for get cost (send variation) API command.

    :param dict obj: Dictionary from API JSON response
    """

    def __init__(self, obj: Dict[str, Any]) -> None:  # noqa: D102
        super().__init__(obj)
        self.__cnt = int(obj.get("cnt", 0))
        self.__cost = float(obj.get("cost", 0.0))

    def __str__(self):
        """Represent object as string."""
        return "<%s count=%d cost=%.2f>" % (
            self.__class__.__name__, self.__cnt, self.__cost)

    def __repr__(self) -> str:
        """Represent object for debug purposes."""
        return str(self)

    @property
    def count(self) -> int:
        """Count of billed message parts."""
        return self.__cnt

    @property
    def cost(self) -> float:
        """Cost of message."""
        return self.__cost


class StatusResponse(Response):
    """
    Response for get status API command.

    :param dict obj: Dictionary from API JSON response
    """

    def __init__(self, obj: Dict[str, Any]) -> None:  # noqa: D102
        super().__init__(obj)
        self.__status = Status(obj)
        self.__data = {}  # type: Dict[str, Any]
        for name in obj.keys():
            self.__data[name] = obj.get(name, None)
        if self.__data["last_date"] is not None:
            self.__data["last_date"] = arrow_get(self.__data["last_date"], "DD.MM.YYYY HH:mm:ss",
                                                 tzinfo=tz.gettz("Europe/Moscow")).datetime
        if self.__data["send_date"] is not None:
            self.__data["send_date"] = arrow_get(self.__data["send_date"], "DD.MM.YYYY HH:mm:ss",
                                                 tzinfo=tz.gettz("Europe/Moscow")).datetime
        for name in ["send_timestamp", "last_timestamp"]:
            del self.__data[name]

    def __str__(self) -> str:
        """Represent object as string."""
        return "<%s id=%d status= %s>" % (self.__class__.__name__, self.__data["id"], self.__status)

    def __repr__(self) -> str:
        """Represent object for debug purposes."""
        return str(self)

    @property
    def status(self) -> Status:
        """Message delivery status with identification."""
        return self.__status

    @property
    def data(self) -> Dict[str, Any]:
        """Delivery status detailed data."""
        return self.__data


class BalanceResponse(Response):
    """
    Response for get account balance API command.

    :param dict obj: Dictionary from API JSON response
    """

    def __init__(self, obj: Dict[str, Any]) -> None:  # noqa: D102
        super().__init__(obj)
        self.__balance = float(obj.get("balance", 0.0))
        self.__credit = float(obj.get("credit", 0.0))
        self.__currency = str(obj.get("currency", ""))

    def __str__(self) -> str:
        """Represent object as string."""
        return "<%s balance=%.2f credit=%.2f currency=%s>" % (
            self.__class__.__name__, self.__balance, self.__credit, self.__currency)

    def __repr__(self) -> str:
        """Represent object for debug purposes."""
        return str(self)

    @property
    def balance(self) -> float:
        """Actual account balance."""
        return self.__balance

    @property
    def credit(self) -> float:
        """Available credit of account (if applied)."""
        return self.__credit

    @property
    def currency(self) -> str:
        """Currency for current account."""
        return self.__currency


smsc/api.py
# -*- coding: utf-8 -*-
"""
smsc.api module.

This module implements the SMSC.ru HTTP API.

:copyright: (c) 2017 by Alexey Shevchenko.
:license: MIT, see LICENSE for more details.
"""
import logging  # noqa: T005
from typing import Any, Dict, List, Optional, Union

import requests  # noqa: T005
from furl import furl

from .exceptions import GetBalanceError, GetCostError, GetStatusError, SendError
from .messages import Message
from .responses import BalanceResponse, CostResponse, SendResponse, StatusResponse

logging.basicConfig(level=logging.DEBUG, format="%(asctime)s {%(pathname)s:%(lineno)d} %(levelname)s - %(message)s")
log = logging.getLogger(__name__)


class SMSC:
    """
    Class for interaction with smsc.ru API.

    Usage::

        >>> from smsc.api import SMSC
        >>> client = SMSC(login="alexey", password="psw")
        >>> client
        <SMSC login='alexey' sender='SMSC.ru'>

    :param str login: Account login name
    :param str password: Password or MD5 hash of password in lower case
    """

    _url = "https://smsc.ru/sys/"

    def __init__(self, login: str, password: str, sender: Optional[str] = None) -> None:  # noqa: D102
        self.__login = login
        self.__password = password
        self.__sender = sender if sender is not None else "SMSC.ru"

    def __str__(self):
        """Represent object as string."""
        return "<%s login='%s' sender='%s'>" % (self.__class__.__name__, self.__login, self.__sender)

    def __repr__(self):
        """Represent object for debug purposes."""
        return str(self)

    @property
    def __auth(self) -> Dict[str, Any]:
        return {"login": self.__login, "psw": self.__password, "fmt": 3}

    def send(self, to: Union[str, List[str]], message: Message) -> SendResponse:
        """
        Send the message.

        Usage::

            >>> from smsc.messages import SMSMessage
            >>> from smsc.api import SMSC
            >>> client = SMSC(login='alexey', password='psw')
            >>> res = client.send(to='79999999999', message=SMSMessage(text='Hello, World!'))  # doctest: +SKIP
            >>> res.count  # doctest: +SKIP
            1
            >>> res.cost  # doctest: +SKIP
            1.44

        :param str|List[str] to: Phone number or list of phone numbers
        :param Message message: Concrete message instance for sending
        :return: Returns the API answer wrapped in the `SendResponse` object
        :rtype: SendResponse
        """
        f = furl(SMSC._url).add(path="send.php").add(self.__auth).add({"sender": self.__sender})
        f.add({"cost": 2, "phones": isinstance(to, str) and to or ",".join(to)})
        f.add(message.encode())
        r = requests.get(f.url)
        if r.status_code != 200:
            raise SendError(str([r.status_code, r.headers, r.text]))
        return SendResponse(r.json())

    def get_cost(self, to: Union[str, List[str]], message: Message) -> CostResponse:
        """
        Retrieve cost of the message.

        Usage::

            >>> from smsc.messages import SMSMessage
            >>> from smsc.api import SMSC
            >>> client = SMSC(login='alexey', password='psw')
            >>> res = client.get_cost(to='79999999999', message=SMSMessage(text='Hello, World!'))  # doctest: +SKIP
            >>> res.count  # doctest: +SKIP
            1
            >>> res.cost  # doctest: +SKIP
            1.44

        :param str|List[str] to: Phone number or list of phone numbers
        :param Message message: Concrete message instance for measure cost
        :return: Returns the API answer wrapped in the `CostResponse` object
        :rtype: CostResponse
        """
        f = furl(SMSC._url).add(path="send.php").add(self.__auth).add({"sender": self.__sender})
        f.add({"cost": 1, "phones": isinstance(to, str) and to or ",".join(to)})
        f.add(message.encode())
        r = requests.get(f.url)
        if r.status_code != 200:
            raise GetCostError(str([r.status_code, r.headers, r.text]))
        return CostResponse(r.json())

    def get_status(self, to: Union[str, List[str]], msg_id: Union[str, List[str]]) -> List[StatusResponse]:
        """
        Get current status of sent message.

        Usage::

            >>> from smsc.api import SMSC
            >>> client = SMSC(login='alexey', password='psw')
            >>> res = client.get_status(to='79999999999', msg_id='1')  # doctest: +SKIP
            >>> res[0].status  # doctest: +SKIP
            <Status status=1 name=Доставлено>

        :param str|List[str] to: Phone number or list of phone numbers
        :param str|List[str] msg_id: Identification of sent message or list of them
        :return: Returns the API answer wrapped in the list of `StatusResponse` objects
        :rtype: List[StatusResponse]
        """
        f = furl(SMSC._url).add(path="status.php").add(self.__auth).add({"charset": "utf-8", "all": 2})
        f.add({"phone": isinstance(to, str) and ",".join([to, ""]) or ",".join(to)})
        f.add({"id": isinstance(msg_id, str) and ",".join([msg_id, ""]) or ",".join(msg_id)})
        r = requests.get(f.url)
        if r.status_code != 200:
            raise GetStatusError(str([r.status_code, r.headers, r.text]))
        res = r.json()
        if isinstance(res, dict):
            raise GetStatusError(str([r.status_code, r.headers, res]))  # pragma: no cover
        result = []
        for obj in res:
            result.append(StatusResponse(obj))
        return result

    def get_balance(self) -> BalanceResponse:
        """
        Get current account balance.

        Usage::

            >>> from smsc.api import SMSC
            >>> client = SMSC(login='alexey', password='psw')
            >>> res = client.get_balance()  # doctest: +SKIP
            >>> res  # doctest: +SKIP
            <BalanceResponse balance=100.01 credit=None currency=RUR>

        :return: Returns the API answer wrapped in the `BalanceResponse` object
        :rtype: BalanceResponse
        """
        f = furl(SMSC._url).add(path="balance.php").add(self.__auth).add({"cur": 1})
        r = requests.get(f.url)
        if r.status_code != 200:
            raise GetBalanceError(str([r.status_code, r.headers, r.text]))
        return BalanceResponse(r.json())
